src/optimizer/query_to_operator_transformer.cpp

```diff
@@ -391,24 +370,117 @@ bool QueryToOperatorTransformer::RequireAggregation(
 
 void QueryToOperatorTransformer::CollectPredicates(
     expression::AbstractExpression *expr) {
+  // First check if all conjunctive predicates are supported before
+  // transfoming
+  // predicate with sub-select into regular predicates
+  std::vector<expression::AbstractExpression *> predicates;
+  util::SplitPredicates(expr, predicates);
+  for (const auto &pred : predicates) {
+    if (!IsSupportedConjunctivePredicate(pred)) {
+      throw Exception("Predicate type not supported yet");
+    }
+  }
+  // Accept will change the expression, e.g. (a in (select b from test)) into
+  // (a IN test.b), after the rewrite, we can extract the table aliases
+  // information correctly
   expr->Accept(this);
   util::ExtractPredicates(expr, predicates_);
 }
 
-bool QueryToOperatorTransformer::GenerateSubquerytree(
-    expression::AbstractExpression *expr,
-    std::vector<expression::AbstractExpression *> &select_list,
-    bool single_join) {
-  if (expr->GetExpressionType() != ExpressionType::ROW_SUBQUERY) {
-    return false;
+bool QueryToOperatorTransformer::IsSupportedConjunctivePredicate(
+    expression::AbstractExpression *expr) {
+  // Currently support : 1. expr without subquery
+  // 2. subquery without disjunction. Since the expr is already one of the
+  // conjunctive exprs, we'll only need to check if the root level is an
+  // operator with subquery
+  if (!expr->HasSubquery()) {
+    return true;
+  }
+  auto expr_type = expr->GetExpressionType();
+  // Subquery with IN
+  if (expr_type == ExpressionType::COMPARE_IN &&
```

Comment:
Should we also check that `expr->GetChild(0) != ExpressionType::ROW_SUBQUERY`? Because [these lines](https://github.com/chenboy/peloton/blob/7066b8c54534c3045a888b985c13fff454332dcc/src/optimizer/query_to_operator_transformer.cpp#L307-L311) wont transform generate the subquery tree for `expr->GetChild(0)`, or is this case handled elsewhere?

Reply:
Nice catch. I've added the check.